### PR TITLE
Fix duplicate micro simulations for macro-points on rank boundaries

### LINKED ISSUE #102: Duplicate micro-simulations on interior boundaries of macro-domain decomposition
Problem:
When using macro-domain decomposition each MPI rank is given a bounding box, which is passed to preCICE to obtain coordinates of coupled points contained therein. If such a coupled macro-point lies on the boundary between two bounding boxes it will be included in both and the Micro Manager currently creates a micro simulation for both duplicates of the macro-point and passes their output to preCICE.
- preCICE interprets this as two contributions to the participant's solution and accumulates them, producing incorrect results.
- duplicate micro simulations receive identical macro-data and with no/global adaptivity identical output, wasting computational effort
- with local adaptivity the two micro-simulations may have different associated active simulations and produce different outputs (differences should be minor if similarity measure is not overly generous)

Suggestion:
The Micro Manager should eliminate duplicate micro simulations to save computational time and avoid incorrect data being received by the macro simulation.

### PR DESCRIPTION
**Pull request by AdityaGupta716**
Closes #102
Problem: When a macro-point lies exactly on the boundary between two rank bounding boxes, preCICE returns it to both ranks. This causes duplicate micro simulations, incorrect accumulated results in preCICE, and wasted computation.
Fix: After get_mesh_vertex_ids_and_coordinates(), each rank uses allgather to see all coords across all ranks, then drops any coord already owned by a lower-ranked rank — ensuring every vertex is processed by exactly one rank while preserving the preCICE ID-coord pairing safely.
Changes:

micro_manager/micro_manager.py: Added deduplication after get_mesh_vertex_ids_and_coordinates()
tests/unit/test_domain_decomposition.py: Removed incorrect test that tested wrong location

### CODE REVIEW

**Comment by AdityaGupta716** [comment]
Hi @IshaanDesai, opening a new PR as suggested. I used allgather instead of gather-scatter because the IDs returned by preCICE are local to each rank and used directly for reading/writing data — redistributing them risked breaking the ID-coord pairing. With allgather, each rank sees all coords and drops any already claimed by a lower-ranked rank, eliminating duplicates safely. Plz review

File: micro_manager/domain_decomposition.py
Comment (IshaanDesai) [inline micro_manager/domain_decomposition.py:1]: Changes in this file do not seem necessary. Please revert if you also think so.

File: micro_manager/micro_manager.py [hunk 1]
```diff
@@ -480,6 +480,33 @@ def initialize(self) -> None:
...
+            # find globally duplicated coords, and remove them from this rank
+            # while preserving the preCICE ID-coord pairing.
+            all_coords = self._comm.allgather(self._mesh_vertex_coords)
+            all_ids = self._comm.allgather(self._mesh_vertex_ids)
+
+            # Build a global set of (coord, id) pairs seen by previous ranks
+            seen_coords = set()
+            keep_mask = np.ones(len(self._mesh_vertex_coords), dtype=bool)
+
+            for r in range(self._size):
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py]:
```suggestion
            for rank in range(self._size):
```

File: micro_manager/micro_manager.py [hunk 2]
```diff
@@ -480,6 +480,33 @@ def initialize(self) -> None:
...
+                for i, coord in enumerate(all_coords[r]):
+                    coord_key = tuple(np.round(coord, decimals=10))
+                    if r < self._rank:
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py]:
```suggestion
                    if rank < self._rank:
```

File: micro_manager/micro_manager.py [hunk 3]
```diff
@@ -480,6 +480,33 @@ def initialize(self) -> None:
...
+                        # Mark coords already claimed by earlier ranks
+                        seen_coords.add(coord_key)
+                    elif r == self._rank:
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py]:
```suggestion
                    elif rank == self._rank:
```

File: tests/unit/test_domain_decomposition.py
Comment (IshaanDesai) [inline tests/unit/test_domain_decomposition.py:1]: Changes in this file are just formatting, please revert.

**Review by IshaanDesai — COMMENTED** [review]
Thanks for the contribution. Please take a look at my comments. Additionally, could you please add a test to ensure that duplicate coordinate filtering is done as expected?

**Comment by AdityaGupta716** [comment]
@IshaanDesai  hi, reverted the changes in domain_decomposition.py as allgather does those fixes at higher level and the formatting-only changes in test_domain_decomposition.py, renamed r to rank in the deduplication loop, and added a new TestDuplicateCoordFiltering test class with three cases — no duplicates, a shared boundary coord between two ranks, and a shared coord across three ranks.

**Comment by IshaanDesai** [comment]
@AdityaGupta716, as a general practice: once you address a review comment by making the relevant changes, you can hit *Resolve conversation*. You don't have to wait for the developer to do this, unless the developer explicitly states so. This makes it easier for the developer, me in this case, to see the changes that are yet to be addressed.

File: tests/unit/test_domain_decomposition.py [hunk 4]
```diff
@@ -95,3 +95,94 @@ def test_rank7_out_of_16_3d(self):
...
         self.assertTrue(np.allclose(mesh_bounds, [0.75, 1, -2, 0, -2, 8]))
+
+
+class TestDuplicateCoordFiltering(TestCase):
+    """
+    Test that duplicate vertex coordinates returned by preCICE on rank boundaries
+    are correctly deduplicated, with lower-ranked ranks taking ownership.
+    """
+
+    def _run_dedup(self, my_rank, size, all_coords, all_ids):
```
Comment (IshaanDesai) [inline tests/unit/test_domain_decomposition.py]: The test should run the actual implementation from `micro_manager.py`, not a copy of it. Take a look at the other tests in this file to see how the functionality can be tested in isolation.

File: tests/unit/test_domain_decomposition.py [hunk 5]
```diff
@@ -95,3 +95,94 @@ def test_rank7_out_of_16_3d(self):
         )
 
         self.assertTrue(np.allclose(mesh_bounds, [0.75, 1, -2, 0, -2, 8]))
+
+
+class TestDuplicateCoordFiltering(TestCase):
+    """
+    Test that duplicate vertex coordinates returned by preCICE on rank boundaries
+    are correctly deduplicated, with lower-ranked ranks taking ownership.
```
Comment (IshaanDesai) [inline tests/unit/test_domain_decomposition.py]:
```suggestion
    are correctly filtered, with lower-ranked ranks taking ownership.
```

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
Good work so far! The implementation looks good, and so do the tests, but the tests should test the actual implementation and not a copy of it. Please take a look at my comments.

**Comment by AdityaGupta716** [comment]
@IshaanDesai Hi done with the changes plz check it out !

**Review by IshaanDesai — COMMENTED** [review]
We are almost there! Can you please make the function `filter_duplicate_coords` a member function of the `DomainDecomposer` class? That would simplify the importing and also future maintenance.

**Comment by AdityaGupta716** [comment]
hi @IshaanDesai done with the changes!

**Review by IshaanDesai — APPROVED** [review]
Great work! Thanks for going through multiple review cycles.
